tests/test_upload.py

```diff
@@ -8,90 +8,79 @@
+        'non_separated_date': ('A1234GY091286', 'A1234GY', date(1986, 12, 9)),
+        'non_separated_long_date': ('A1234GY09121986', 'A1234GY', date(1986, 12, 9)),
+        'space_separated_date': ('A1234GY 09 12 86', 'A1234GY', date(1986, 12, 9)),
+        'non_zero_padded_date': ('A1234GY 9/6/86', 'A1234GY', date(1986, 6, 9)),
+        'four_digit_year': ('A1234GY 09/12/1986', 'A1234GY', date(1986, 12, 9)),
+        'reference_generator_format': ('A1234GY/09/12/1986', 'A1234GY', date(1986, 12, 9)),
+
+        'reversed_format_1': ('09/12/86 A1234GY', 'A1234GY', date(1986, 12, 9)),
+        'reversed_format_2': ('10/06/80 A1214GY', 'A1214GY', date(1980, 6, 10)),
+        'reversed_with_whitespace': ('  09/12/86 A1234GY  ', 'A1234GY', date(1986, 12, 9)),
+        'reversed_with_whitespace_long_year': ('  09/12/1986 A1234GY  ', 'A1234GY', date(1986, 12, 9)),
+        'reversed_with_separators': ('9/12/86/A1234GY', 'A1234GY', date(1986, 12, 9)),
+        'reversed_with_separators_long_year': ('9/12/1986/A1234GY', 'A1234GY', date(1986, 12, 9)),
+        'reversed_without_spaces': ('091286A1234GY', 'A1234GY', date(1986, 12, 9)),
+        'reversed_without_spaces_long_year': ('09121986A1234GY', 'A1234GY', date(1986, 12, 9)),
+    }
+    unsuccessful = {
+        'blank': '',
+        'null': None,
+        'invalid': 'JOHN HALLS',
+        'no_date': 'A1234GY',
+        'no_prisoner_number': '09/12/1986',
+
+        'truncated_prisoner_number': 'A1234Y 09/12/1986',
+        'long_prisoner_number': 'AA1234GY 09/12/1986',
+        'one_digit_year': 'A1234GY 1/1/1',
+        'three_digit_year': 'A1234GY 09/12/198',
+        'too_many_date_digits': 'A1234GY 0931231986',
+        'trailing_digits': 'A1234GY 09/12/198600000',
+        'impossible_day': 'A1234GY 32/12/1986',
+        'impossible_month': 'A1234GY 09/13/1986',
+        'invalid_leap_day': 'A1234GY 29/02/1981',
+
+        'reversed_trailing_letters': '29/02/1981 A1234GYA',
+        'reversed_invalid_leap_day': '29/02/1981 A1234GY',
+    }
+
+    @classmethod
+    def create_tests(cls):
+        for name, values in cls.successful.items():
+            cls.add_successful_method(name, *values)
+        for name, reference in cls.unsuccessful.items():
+            cls.add_unsuccessful_method(name, reference)
+
+    @classmethod
+    def add_successful_method(cls, name, reference, prisoner_number, prisoner_dob):
+        def method(self):
+            parsed_number, parsed_dob = upload.parse_credit_reference(reference)
+            self.assertEqual(parsed_number, prisoner_number)
+            self.assertEqual(parsed_dob, prisoner_dob)
+
+        setattr(cls, 'test_%s_parses' % name, method)
+
+    @classmethod
+    def add_unsuccessful_method(cls, name, reference):
+        def method(self):
+            self.assertEqual(upload.parse_credit_reference(reference), None)
+
+        setattr(cls, 'test_%s_does_not_parse' % name, method)
+
+
+CreditReferenceParsingTestCase.create_tests()
```

Comment:
it makes it easier to add additional failing and passing references and the tables of expected results are no less readable than before… i don't think there's any benefit to verbosely writing out every line as an individual method